### PR TITLE
[GH-507,699] Add Support for DM subscriptions for watchers and reporter

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
**Summary** 

- In order to get a DM notification from the Jira plugin, the user has to meet certain criteria like being assigned or mentioned. But now even if a user is a reporter or watching that issue will also get a DM notification.
- This PR contains all the changes from [Pr #840](https://github.com/mattermost/mattermost-plugin-jira/pull/840) So this PR alone is enough to fix this issue so we can close [Pr #840](https://github.com/mattermost/mattermost-plugin-jira/pull/840) afterward.

**Issue** 

- Fixes https://github.com/mattermost/mattermost-plugin-jira/issues/699
- Fixes https://github.com/mattermost/mattermost-plugin-jira/issues/507
- [Pr #840](https://github.com/mattermost/mattermost-plugin-jira/pull/840)

### CODE REVIEW

File: server/settings.go [hunk 1]
```diff
@@ -31,21 +40,48 @@ func (p *Plugin) settingsNotifications(header *model.CommandArgs, instanceID, ma
 	if connection.Settings == nil {
 		connection.Settings = &ConnectionSettings{}
 	}
-	connection.Settings.Notifications = value
+	switch args[1] {
+	case subCommandAssignee:
```
Comment (javaguirre) [inline server/settings.go:44]: WDYT about having a:

```go
connection.Settings.SendNotificationsFor(role, &value)
```

Then we wouldn't need these long switches and I think it would be easier to maintain? Let me know your thoughts!

File: server/settings.go [hunk 2]
```diff
@@ -31,21 +40,48 @@ func (p *Plugin) settingsNotifications(header *model.CommandArgs, instanceID, ma
...
 	updatedConnection, err := p.userStore.LoadConnection(instanceID, mattermostUserID)
 	if err != nil {
-		return p.responsef(header, "Your username is not connected to Jira. Please type `jira connect`. %v", err)
+		return p.responsef(header, errConnectToJira, err)
 	}
 	notifications := settingOff
-	if updatedConnection.Settings.Notifications {
-		notifications = settingOn
+	switch args[1] {
+	case subCommandAssignee:
```
Comment (javaguirre) [inline server/settings.go:68]: Idem here, something like:

```go
settings := *updatedConnection.Settings

... 

if settings.NeedToSendNotificationsFor(role) { ... }
```

we would translate the subcommandRole to the role, but we can use a Golang map to do that (for example).

WDYT? Thank you! :-)

File: server/webhook.go [hunk 3]
```diff
@@ -137,7 +138,29 @@ func (wh *webhook) PostNotifications(p *Plugin, instanceID types.ID) ([]*model.P
 		}
 		// If this is a comment-related webhook, we need to check if they have permissions to read that.
 		// Otherwise, check if they can view the issue.
-
+		switch notification.notificationType {
+		case subCommandAssignee:
```
Comment (javaguirre) [inline server/webhook.go:142]: I think we could follow the same idea here, having something in common with all these switch case statements we could handle it in the settings and through parameters to the function.

**Review by javaguirre — COMMENTED** [review]
Hi! I'm Javi, from Mattermost (first week), I have a couple of comments, let me know what you think!

Thank you for the hard work! This is a tricky one! :-)

File: server/settings.go (hunk 1 above)
Comment (javaguirre) [reply server/settings.go:44]: The problem is not the switch case itself in my opinion, but the changes in `ConnectionSettings` to support the notifications, let me show you what I mean tomorrow. I think we could try to make it easier to handle.

**Comment by javaguirre** [comment]
@Kshitij-Katiyar @Nityanand13 This is what I mean, I think changing the way we store the data in the struct would make us do much less code and this code would be easier to test.

https://replit.com/@javaguirre/GH-507699#main.go

Let me know what you think!

**Review by javaguirre — CHANGES_REQUESTED** [review]
Let me know when you have an opinion on my approach! Thank you! :-)

**Comment by Nityanand13** [comment]
> 

@javaguirre Sorry for the late reply. Your approach seems good and now I am implementing the same.

**Comment by javaguirre** [comment]
Thank you @Nityanand13 ! Let me know if you need anything! 👍

File: server/settings.go (hunk 2 above)
Comment (Nityanand13) [reply server/settings.go]: It seems better and now I have implemented the same.

File: server/webhook.go (hunk 3 above)
Comment (Nityanand13) [reply server/webhook.go]: Yes, modified this as well.

File: server/user.go [hunk 4]
```diff
@@ -43,15 +43,33 @@ func (c *Connection) JiraAccountID() types.ID {
 }
 
 type ConnectionSettings struct {
-	Notifications bool `json:"notifications"`
+	Notifications          bool `json:"notifications"`
+	RolesForDMNotification map[string]*bool
```
Comment (javaguirre) [inline server/user.go:47]: Thank you for the changes! I think we need much less code now. 

I am wondering, do you think we need to store space (using pointers here) for booleans? wouldn't you think would be better having only bool?

File: server/issue.go [hunk 5]
```diff
@@ -1039,3 +1043,154 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+		if err != nil {
+			continue
+		}
+
+		client, err := instance.GetClient(connection)
+		if err != nil {
+			continue
+		}
+
+		return client, connection, nil
```
Comment (javaguirre) [inline server/issue.go:1192]: I think we could get this out to a function, so later we could unittest it easier, I believe it's a small change, WDYT?

Thank you!

File: server/issue.go [hunk 6]
```diff
@@ -1039,3 +1043,154 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+			"AccountID": wh.Issue.Fields.Reporter.AccountID,
+			"Name":      wh.Issue.Fields.Reporter.Name,
+		})
+	}
+
+	instance, err := p.instanceStore.LoadInstance(instanceID)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, account := range accountInformation {
```
Comment (javaguirre) [inline server/issue.go:1171]: I'm a bit scared about not having testing for these, do you think we could spend a bit of time getting come test coverage here so we spot early on problems that may arise?

Thank you!

**Review by javaguirre — CHANGES_REQUESTED** [review]
Thank you! I really like the changes, I think it's much cleaner now.

I have some details to finish, let me know what you think!

File: server/issue.go (hunk 6 above)
Comment (javaguirre) [reply server/issue.go:1171]: I mean these new methods for webhooks, I thought It would highlight all the lines, but it didn't. :-)

File: server/issue.go (hunk 5 above)
Comment (javaguirre) [reply server/issue.go:1192]: I mean the code inside the for loop

**Comment by javaguirre** [comment]
Could you post a video on how to use this new change? Something like this so it's easier to QA.

https://github.com/mattermost/mattermost-plugin-jira/pull/863#issuecomment-1200475963

Thank you!

**Comment by Nityanand13** [comment]
> Could you post a video on how to use this new change? Something like this so it's easier to QA.
> 
> [#863 (comment)](https://github.com/mattermost/mattermost-plugin-jira/pull/863#issuecomment-1200475963)
> 
> Thank you!

Vedio link for this PR:
1. https://drive.google.com/file/d/1lSTDElqTwI1_XTwunTtaSCepWroMyb7Z/view?usp=sharing
2. https://drive.google.com/file/d/1Xpi9jdftdPk0NBLZ8yROJmjMLhKRuK82/view?usp=sharing

**Comment by javaguirre** [comment]
Thank you very much @Nityanand13 ! Let me know when testing is done @Kshitij-Katiyar . :+1:

**Comment by javaguirre** [comment]
@Nityanand13 Can't open the videos in any of the PRs, can you check the extension of the files and tell me what kind of video is so I can open it? Also If I could already open it online would be nice, with something like loom or whatever video Drive supports.

Thank you very much!

**Comment by Nityanand13** [comment]
> @Nityanand13 Can't open the videos in any of the PRs, can you check the extension of the files and tell me what kind of video is so I can open it? Also If I could already open it online would be nice, with something like loom or whatever video Drive supports.
> 
> Thank you very much!

Vedio link for this PR:
1. [Add_Support_for_DM_subscriptions_for_watchers.webm](https://user-images.githubusercontent.com/107465508/194276591-3d30064d-9839-4a97-b03f-83ffb58a3feb.webm)

2. [Add_Support_for_DM_subscriptions_for_reporter.webm](https://user-images.githubusercontent.com/107465508/194276979-0c725621-0688-4c45-8593-e6cdb5839cdf.webm)

**Comment by javaguirre** [comment]
The CI is failing @Nityanand13 , let me know if you can check it out! Thank you!

**Comment by javaguirre** [comment]
I see the changes are in the other PR

File: server/issue.go [hunk 7]
```diff
@@ -518,24 +521,25 @@ func (p *Plugin) httpGetJiraProjectMetadata(w http.ResponseWriter, r *http.Reque
...
+		issueTypes := make([]option, len(project.IssueTypes))
+		for index, issueType := range project.IssueTypes {
+			if issueType.Subtasks {
 				continue
 			}
-			issueTypes = append(issueTypes, utils.ReactSelectOption{
-				Value: issue.ID,
-				Label: issue.Name,
-			})
+			issueTypes[index] = option{
```
Comment (mickmister) [inline server/issue.go:537]: Are these changes relevant to the purpose of this PR? Also it seems we may introduce a bug here, since the `continue` statement above will result in a null pointer on the frontend, since we are "skipping" over that index.

File: server/settings.go [hunk 8]
```diff
@@ -9,17 +11,34 @@ import (
...
+	errStoreNewSettings = "Could not store new settings. Please contact your system administrator. Error: %v"
+	errConnectToJira    = "Your account is not connected to Jira. Please type `/jira connect`. %v"
+	subCommandAssignee  = "assignee"
+	subCommandMention   = "mention"
+	subCommandReporter  = "reporter"
+	subCommandWatching  = "watching"
 )
 
+func (connection *Connection) sendNotification(role string, hasNotification bool) bool {
+	if role != subCommandAssignee && role != subCommandMention && role != subCommandReporter && role != subCommandWatching {
```
Comment (mickmister) [inline server/settings.go:24]: Since these aren't referring to actual subcommands here, I recommend using different names for these constants. Maybe `notificationRoleAssignee` etc.

File: server/settings.go [hunk 9]
```diff
@@ -31,21 +50,24 @@ func (p *Plugin) settingsNotifications(header *model.CommandArgs, instanceID, ma
 	if connection.Settings == nil {
 		connection.Settings = &ConnectionSettings{}
 	}
-	connection.Settings.Notifications = value
+	if !connection.sendNotification(args[1], value) {
+		return p.responsef(header, helpText)
+	}
```
Comment (mickmister) [inline server/settings.go:55]: It's a little confusing how `sendNotification` mutates the setting state, and is at the same time used as a boolean control mechanism here. Also the name `sendNotification` sounds like it should be sending a notification, but that's not what's happening here

File: server/subscribe_test.go [hunk 10]
```diff
@@ -1381,7 +1381,6 @@ func TestGetChannelsSubscribed(t *testing.T) {
 			r := bytes.NewReader(data)
 			bb, err := ioutil.ReadAll(r)
 			require.Nil(t, err)
-
```
Comment (mickmister) [inline server/subscribe_test.go:1384]: Can you please remove this change, to keep the git history of the file more appropriate?

File: server/webhook_worker.go [hunk 11]
```diff
@@ -39,15 +39,16 @@ func (ww webhookWorker) process(msg *webhookMessage) (err error) {
...
-
-	if _, _, err = wh.PostNotifications(ww.p, msg.InstanceID); err != nil {
-		ww.p.errorf("WebhookWorker id: %d, error posting notifications, err: %v", ww.id, err)
-	}
-
 	v := wh.(*webhook)
 	if err = v.JiraWebhook.expandIssue(ww.p, msg.InstanceID); err != nil {
 		return err
 	}
+	ww.p.checkIssueWatchers(v, msg.InstanceID)
```
Comment (mickmister) [inline server/webhook_worker.go:46]: Can you please add a newline before this line?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work on this @Kshitij-Katiyar! have a just a few questions here, and a request to remove changes unrelated to the purpose of the PR. Please let me know what you think :+1:

**Comment by javaguirre** [comment]
@Kshitij-Katiyar @Nityanand13 Is this ready to be reviewed again? Please request the review again so I effectively know I have to review again and we don't lose time, thank you!

File: server/settings.go [hunk 12]
```diff
@@ -2,24 +2,44 @@ package main
...
+	const helpText = "`/jira settings notifications [assignee|mention|reporter|watching] [value]`\n* Invalid value. Accepted values are: `on` or `off`."
 
-	if len(args) != 2 {
+	if len(args) != 3 {
 		return p.responsef(header, helpText)
 	}
 
 	var value bool
-	switch args[1] {
+	switch args[2] {
```
Comment (javaguirre) [inline server/settings.go:42]: These magic numbers would be much better in a constant, so they're named in a descriptive way.

https://hireamir.com/blog/software-anti-patterns-magic-numbers

File: server/issue.go [hunk 13]
```diff
@@ -1039,3 +1043,164 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+		return nil, connection, err
+	}
+
+	return client, connection, nil
+}
+
+func (wh *webhook) fetchConnectedUser(p *Plugin, instanceID types.ID) (Client, *Connection, error) {
+	var accountInformation []map[string]string
+
+	if wh.Issue.Fields != nil && wh.Issue.Fields.Creator != nil {
```
Comment (javaguirre) [inline server/issue.go:1171]: These is repeated three times, why not a function?

File: server/issue.go [hunk 14]
```diff
@@ -1039,3 +1043,164 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+
+	// Check old setting for backwards compatibility
+	return s.Notifications
+}
+
+func (p *Plugin) fetchConnectedUserFromAccount(account map[string]string, instance Instance) (Client, *Connection, error) {
+	var mattermostUserID types.ID
+	var err error
+	if account["AccountID"] != "" {
+		mattermostUserID, err = p.userStore.LoadMattermostUserID(instance.GetID(), account["AccountID"])
```
Comment (javaguirre) [inline server/issue.go:1147]: if only the account key is changing, the rest of the code can be done after the if to improve readability.

```go
accountKey := account["Name"]

if account["AccountID"] ... {
  accountKey = ...
} 

p.userStore....

```

**Review by javaguirre — CHANGES_REQUESTED** [review]

File: server/issue.go [hunk 15]
```diff
@@ -1039,3 +1045,160 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+	})
+}
+
+func (p *Plugin) GetUserSetting(wh *webhook, instanceID types.ID, jiraAccountID, jiraUsername string) (*Connection, error) {
+	instance, err := p.instanceStore.LoadInstance(instanceID)
+	if err != nil {
+		return nil, err
+	}
+	var mattermostUserID types.ID
+	if jiraAccountID != "" {
```
Comment (javaguirre) [inline server/issue.go:1119]: Only a variable assignment depends on a condition, the rest should be out.

```golang
jiraUserName := jiraAccountID

if jiraUserName != "" {
  jiraUserName = jiraUsername
}

mattermostUserID, err = p.userStore.LoadMattermostUserID(...)

```

File: server/issue.go [hunk 16]
```diff
@@ -1039,3 +1045,160 @@ func (p *Plugin) getClient(instanceID, mattermostUserID types.ID) (Client, Insta
...
+}
+
+func (p *Plugin) applyReporterNotification(wh *webhook, instanceID types.ID, reporter *jira.User) {
+	if !wh.eventTypes.ContainsAny(createdCommentEvent) {
+		return
+	}
+
+	jwhook := wh.JiraWebhook
+	if reporter == nil ||
+		(reporter.Name != "" && reporter.Name == jwhook.User.Name) ||
```
Comment (javaguirre) [inline server/issue.go:1090]: can `jwhook.User.Name` be `""`?

**Review by javaguirre — CHANGES_REQUESTED** [review]
Only a small detail

File: server/issue.go (hunk 16 above)
Comment (Nityanand13) [reply server/issue.go:1090]: Yes

**Comment by mickmister** [comment]
@Kshitij-Katiyar Is this ready for re-review?

**Comment by Nityanand13** [comment]
> @Kshitij-Katiyar Is this ready for re-review?

Yes @mickmister

File: server/command_test.go [hunk 17]
```diff
@@ -165,12 +173,12 @@ func TestPlugin_ExecuteCommand_Settings(t *testing.T) {
...
 			commandArgs:  &model.CommandArgs{Command: "/jira settings", UserId: mockUserIDWithNotifications},
 			numInstances: 1,
-			expectedMsg:  "Current settings:\n\tNotifications: on",
+			expectedMsg:  "Current settings:\n\t- Notifications for assignee: on \n\t- Notifications for mention: on \n\t- Notifications for reporter: on \n\t- Notifications for watching: on",
 		},
 		"no params, without notifications": {
 			commandArgs:  &model.CommandArgs{Command: "/jira settings", UserId: mockUserIDWithoutNotifications},
 			numInstances: 1,
-			expectedMsg:  "Current settings:\n\tNotifications: off",
+			expectedMsg:  "Current settings:\n\t- Notifications for assignee: off \n\t- Notifications for mention: off \n\t- Notifications for reporter: off \n\t- Notifications for watching: off",
```
Comment (mickmister) [inline server/command_test.go:181]: Unfortunate how we're not asserting the actual changes being done to the user settings here. This is probably fine though.

File: server/issue_test.go [hunk 18]
```diff
@@ -279,12 +269,214 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			totalNotifications := len(wh.notifications)
+			p.applyReporterNotification(wh, tt.instanceID, tt.reporter)
+			if tt.reporter == nil || tt.instanceID == "test-instanceID" {
+				assert.Equal(t, len(wh.notifications), totalNotifications)
+			} else {
+				assert.Equal(t, len(wh.notifications), 1+totalNotifications)
+			}
```
Comment (mickmister) [inline server/issue_test.go:427]: Instead of having if statements that have implicit knowledge about the test data, the individual test cases should provide enough context to make it so there's no logic here. Such as having a string field `expectedError` that contains the error we expect to receive for that test case, and is empty when we expect no error.

File: server/issue_test.go [hunk 19]
```diff
@@ -279,12 +269,214 @@ func TestRouteShareIssuePublicly(t *testing.T) {
...
+		watchingRole: {
+			role:         watchingRole,
+			notification: false,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			val := cs.ShouldReceiveNotification(tt.role)
+			assert.Equal(t, tt.notification, val)
+		})
+	}
```
Comment (mickmister) [inline server/issue_test.go:302]: For completeness, we should also test the backwards compability case

**Review by mickmister — APPROVED** [review]
No blocking issues found. Excellent work on this @Kshitij-Katiyar :+1: This feature will deliver a lot of value.

**Comment by mickmister** [comment]
@DHaussermann Given the high impact of this feature, I'm wondering what's the bare minimum we need to do here to get an RC on community with this change

**Comment by mickmister** [comment]
@DHaussermann On second thought, are we supposed to avoid shipping features like this right now? cc @m1lt0n

**Comment by m1lt0n** [comment]
@mickmister cc: @DHaussermann I believe so Michael, since it's not strictly a bug fix, but a change in the business logic. We can discuss sync during our planning calls.

**Comment by hanzei** [comment]
@Kshitij-Katiyar Heads up that there are conflicts to resolve

File: server/issue_test.go (hunk 19 above)
Comment (Nityanand13) [reply server/issue_test.go:302]: @mickmister Can you please elaborate on what you mean by `test the backward compatibility case`?

As earlier, in order to get a DM notification, the user has to meet certain criteria like being assigned or mentioned. But now we added support for DM subscriptions for the watchers and reporters as well.

Comment (mickmister) [reply server/issue_test.go:302]: @Nityanand13 I mean test the case where `settings.Notifications` is `true` or `false`, and the rest of the user notification settings are not defined yet, because that's going to be the case for every current installation of the plugin

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar Heads up that there are conflicts to resolve

@hanzei Fixed merge conflicts

File: server/command.go [hunk 20]
```diff
@@ -252,12 +252,29 @@ func createSettingsCommand(optInstance bool) *model.AutocompleteData {
...
+	assigneeNotifications.AddStaticListArgument("value", true, setting)
+
+	mentionNotifications := model.NewAutocompleteData(mentionRole, "", "manage mention notifications")
+	mentionNotifications.AddStaticListArgument("value", true, setting)
+
+	reporterNotifications := model.NewAutocompleteData(reporterRole, "", "manage reporter notifications")
+	reporterNotifications.AddStaticListArgument("value", true, setting)
+
+	watchingNotifications := model.NewAutocompleteData(watchingRole, "", "manage watching notifications")
+	reporterNotifications.AddStaticListArgument("value", true, setting)
```
Comment (srkgupta) [inline server/command.go:273]: @Kshitij-Katiyar Can we also update the help Text in the file `server/plugin/template.go`. 

Currently the help text states:

```
/jira instance settings [setting] [value] - Update your user settings
[setting] can be notifications
[value] can be on or off
```

This might also need some update as we have now included additional options.

**Review by srkgupta — APPROVED** [review]
Overall LGTM. Just left one minor suggestion.

File: server/command.go (hunk 20 above)
Comment (Nityanand13) [reply server/command.go:274]: @srkgupta I have updated the `commonHelpText` in `server/command.go`. 
Can you please elaborate what do you mean by 
> Can we also update the help Text in the file `server/plugin/template.go` ?

as we do not have any such file.

**Comment by tomcat78** [comment]
Hey, is there still something blocking this issue? I'd love to have the option to get notifications on tickets, that I watch.

**Comment by mickmister** [comment]
@Kshitij-Katiyar Can you please resolve the merge conflicts here? Thank you :+1:

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar Can you please resolve the merge conflicts here? Thank you 👍

@mickmister sure resolving them.

**Comment by mickmister** [comment]
@Dhaussermann This PR is ready for QA review

**Comment by mickmister** [comment]
@Kshitij-Katiyar I think we should tell the user _why_ they received a notification. e.g. "because you are watching x" or "because you reported x"

**Comment by hanzei** [comment]
@Kshitij-Katiyar Heads up that there is a conflict to resolve.

**Comment by mickmister** [comment]
> @Kshitij-Katiyar I think we should tell the user _why_ they received a notification. e.g. "because you are watching x" or "because you reported x"

I'd like to reiterate this point. I think this would be a big help when using this feature

**Comment by Kshitij-Katiyar** [comment]
This PR was very outdated, which is why we had to create a new one. All the changes from this PR are included in https://github.com/mattermost/mattermost-plugin-jira/pull/1097, so we are closing this one.
